modules/detection/TrackVADEmitter.js

```diff
@@ -0,0 +1,237 @@
+        // We don't need stereo for determining the VAD score so we create a single channel processing node.
+        this._audioProcessingNode = this._audioContext.createScriptProcessor(this._procNodeSampleRate, 1, 1);
+    }
+
+    /**
+     * ScriptProcessorNode callback, the input parameters contains the PCM audio that is then sent to rnnoise.
+     * Rnnoise only accepts PCM samples of 480 bytes whereas the webaudio processor node can't sample at a multiple
+     * of 480 thus after each _onAudioProcess callback there will remain and PCM buffer residue equal
+     * to _procNodeSampleRate / 480 which will be added to the next sample buffer and so on.\
+     *
+     *
+     * @param {AudioProcessingEvent} audioEvent - Audio event.
+     * @returns {void}
+     * @fires VAD_SCORE_PUBLISHED
+     */
+    _onAudioProcess(audioEvent) {
+        // Prepend the residue PCM buffer from the previous process callback.
+        const inData = audioEvent.inputBuffer.getChannelData(0);
+        const completeInData = [ ...this._bufferResidue, ...inData ];
+        const sampleTimestamp = Date.now();
+
+        let i = 0;
+
+        for (; i + this._vadSampleSize < completeInData.length; i += this._vadSampleSize) {
+            const pcmSample = completeInData.slice(i, i + this._vadSampleSize);
+            const vadScore = this._vadProcessor.calculateAudioFrameVAD(pcmSample);
+
+            /**
+             * VAD score publish event
+             *
+             * @event VAD_SCORE_PUBLISHED
+             * @type {Object}
+             * @property {Date}   timestamp - Exact time at which processed PCM sample was generated.
+             * @property {number} score - VAD score on a scale from 0 to 1 (i.e. 0.7)
+             * @property {string} deviceId - Device id of the associated track.
+             */
+            this.emit(VAD_SCORE_PUBLISHED, {
+                timestamp: sampleTimestamp,
+                score: vadScore,
+                deviceId: this._localTrack.getDeviceId()
+            });
+        }
+
+        this._bufferResidue = completeInData.slice(i, completeInData.length);
+    }
+
+    /**
+     * Connects the nodes in the AudioContext to start the flow of audio data.
+     *
+     * @returns {void}
+     */
+    _connectAudioGraph() {
+        this._audioProcessingNode.onaudioprocess = this._onAudioProcess;
+        this._audioSource.connect(this._audioProcessingNode);
+        this._audioProcessingNode.connect(this._audioContext.destination);
+    }
+
+    /**
+     * Disconnects the nodes in the AudioContext.
+     *
+     * @returns {void}
+     */
+    _disconnectAudioGraph() {
+        // Even thought we disconnect the processing node it seems that some callbacks remain queued,
+        // resulting in calls with and uninitialized context.
+        // eslint-disable-next-line no-empty-function
+        this._audioProcessingNode.onaudioprocess = () => {};
```

Comment:
Noup I just saw this approach being used in other parts of jitsi so I assumed it's the de facto way of doing it.

Reply:
Not really. I was just wondering if it would be more optimal with undefined since you mentioned the function is called even after the processing node is disconnected...